_posts/2024-01-26-cats-effect-parallelism.md

```diff
@@ -0,0 +1,789 @@
+  
+First of all, we will need to define a basic route which will let us accept organization name as a query parameter, so let's add a new case for our `routes` definition: 
+
+```scala
+def routes: HttpRoutes[IO] = {
+  HttpRoutes.of[IO] { 
+    case GET -> Root => Ok("hi :)")
+    case GET -> Root / "org" / orgName => Ok(orgName)
+  }
+}
+```
+
+Let's quickly test the new route and then proceed with the important stuff.
+
+![alt ""](../images/github-contributors-aggregator/test_basic_route.png)
+
+Great!
+
+Here's a list of things we need to prepare before writing the main logic:
+- handling GitHub Token (configuration, loading, attaching to request headers)
+- mechanism to create proper GitHub REST API URL-s (parameterization, pagination)
+- general approach for issuing a single HTTP request (deserialization, error handling)
+
+As soon as we have those components, we can start composing our main piece of logic.
+
+Since this is a small project, it is not required to overwhelm ourselves with the approaches that quickly can turn into the rabbit hole of overengineering practices.
+
+I suggest that we create a simple private methods in `Main.scala` which will be just called multiple times inside `routes` definition.
+
+### 7.1 Handling Github Token
+
+Github token is just a string which must be attached to request headers so that Github REST API rate limiter doesn't prohibit us issuing huge amount of requests. 
+
+Let's be simple and model that as a single field case class:
+
+```scala
+object Main extends IOApp.Simple {
+
+  final case class Token(token: String) derives ConfigReader {
+    override def toString: String = token
+  }
+
+  /**
+   * other code
+   */
+}
+```
+
+We've defined a case class named `Token` which derives from `ConfigReader` suggesting that it may be used in some configuration reading context.
+
+The derives clause is part of the new Scala 3 feature called "derivation," which allows automatic generation of type class instances for case classes and other data types. In this case, it generates a `ConfigReader` instance for our `Token` class, allowing us to read instances of `Token` from some configuration source.
+
+In general, there are a few options for managing such things:
+- configuration
+  - easy to set up
+- env variable 
+  - easy to set up
+- key chains or secret management systems
+  - not so easy and requires some time to set up, definitely an overkill for a small project such as ours
+- hardcoding
+  - very good for prototyping and development but extremely unsafe, as soon as you publish your token, GitHub will immediately expire it so that nobody can abuse it on your behalf
+
+Since I've already showed you `ConfigReader` and `pureconfig` dependency in our `build.sbt` it's fairly easy to guess which option we're using in this project, just a personal preference.
```

Comment:
Nit: try to avoid "I" and "you" - this gives the impression that the article talks _at_ the reader instead of _with_ the reader. 

Best to use "together" words. For example here, "I've already showed you" can be "we've already seen".

Reply:
Okay, thanks for advice, will have a look in different places too